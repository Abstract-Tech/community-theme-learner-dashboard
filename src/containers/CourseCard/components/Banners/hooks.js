import { useIntl } from '@edx/frontend-platform/i18n';
import { useCardValues } from 'hooks';
import { selectors } from 'data/redux';

const { cardData } = selectors;

export const useCourseBannerData = ({ courseNumber }) => ({
  courseData: useCardValues(courseNumber, {
    isVerified: cardData.isVerified,
    isCourseRunActive: cardData.isCourseRunActive,
    canUpgrade: cardData.canUpgrade,
    isAuditAccessExpired: cardData.isAuditAccessExpired,
    courseWebsite: cardData.courseWebsite,
  }),
  formatMessage: useIntl().formatMessage,
});

export default {
  useCourseBannerData,
};
