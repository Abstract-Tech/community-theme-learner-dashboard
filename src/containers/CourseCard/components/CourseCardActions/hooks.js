import { useDispatch } from 'react-redux';

import { Locked } from '@edx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import { hooks as appHooks } from 'data/redux';
import messages from './messages';

export const useCardActionData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const dispatch = useDispatch();
  const {
    canUpgrade,
    coursewareAccess,
    hasStarted,
    isAudit,
    isAuditAccessExpired,
    isVerified,
  } = appHooks.useCardEnrollmentData(cardId);
  const { isArchived } = appHooks.useCardCourseRunData(cardId);
  const {
    canChange,
    hasSessions,
    isEntitlement,
    isExpired,
    isFulfilled,
  } = appHooks.useCardEntitlementData(cardId);
  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(dispatch, cardId);

  const hasAccess = coursewareAccess.isStaff || !(
    coursewareAccess.hasUnmetPrereqs
    || coursewareAccess.isTooEarly
    || isArchived);

  const selectSessionButton = {
    children: formatMessage(messages.selectSession),
    onClick: openSessionModal,
    disabled: !hasAccess || (!canChange || !hasSessions),
  };

  const viewCourseButton = {
    children: formatMessage(messages.viewCourse),
    disabled: !hasAccess || (isEntitlement && isExpired),
  };

  const beginCourseButton = {
    children: formatMessage(messages.beginCourse),
    disabled: !hasAccess,
  };

  const resumeButton = {
    children: formatMessage(messages.resume),
    disabled: isAudit && isAuditAccessExpired,
  };

  const upgradeButton = {
    iconBefore: Locked,
    variant: 'outline-primary',
    children: formatMessage(messages.upgrade),
    disabled: !canUpgrade,
  };

  let primary;
  if (isEntitlement) {
    primary = isFulfilled ? selectSessionButton : viewCourseButton;
  } else if (!hasStarted) {
    primary = beginCourseButton;
  } else if (isArchived) {
    primary = viewCourseButton;
  } else {
    primary = resumeButton;
  }

  const secondary = (isEntitlement || !isVerified) ? null : upgradeButton;
  return { primary, secondary };
};

export default useCardActionData;
