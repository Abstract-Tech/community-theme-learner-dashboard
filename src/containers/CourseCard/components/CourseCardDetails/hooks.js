import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

import * as module from './hooks';
import messages from './messages';

export const useAccessMessage = ({ cardId }) => {
  const { formatMessage, formatDate } = useIntl();
  const enrollment = appHooks.useCardEnrollmentData(cardId);
  const courseRun = appHooks.useCardCourseRunData(cardId);
  if (!courseRun.isStarted) {
    return formatMessage(messages.courseStarts, { startDate: courseRun.startDate });
  }
  if (enrollment.isEnrolled) {
    if (enrollment.isAudit) {
      const {
        accessExpirationDate,
        isAuditAccessExpired,
      } = enrollment;
      return formatMessage(
        isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
        { accessExpirationDate: formatDate(accessExpirationDate) },
      );
    }
    const { isArchived, endDate } = courseRun;
    return formatMessage(
      isArchived ? messages.courseEnded : messages.courseEnds,
      { endDate: formatDate(endDate) },
    );
  }
  return null;
};

export const useCardDetailsData = ({ dispatch, cardId }) => {
  const { formatMessage } = useIntl();
  const providerName = appHooks.useCardProviderData(cardId).name;
  const { courseNumber } = appHooks.useCardCourseData(cardId);
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = appHooks.useCardEntitlementsData(cardId);

  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(dispatch, cardId);

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: module.useAccessMessage({ cardId }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    formatMessage,
    courseNumber,
  };
};

export default useCardDetailsData;
