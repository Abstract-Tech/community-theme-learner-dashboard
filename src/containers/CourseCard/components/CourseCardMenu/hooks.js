import React from 'react';
import { StrictDict } from 'utils';
import * as module from './hooks';

export const state = StrictDict({
  isUnenrollConfirmVisible: (val) => React.useState(val),
  isEmailSettingsVisible: (val) => React.useState(val),
});

export const unenrollModalHooks = () => {
  const [isVisible, setIsVisible] = module.state.isUnenrollConfirmVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const emailSettingsModalHooks = () => {
  const [isVisible, setIsVisible] = module.state.isEmailSettingsVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const menuHooks = () => {
  const unenrollModal = module.unenrollModalHooks();
  const emailSettingsModal = module.emailSettingsModalHooks();
  return {
    emailSettingsModal,
    unenrollModal,
  };
};

export default menuHooks;
