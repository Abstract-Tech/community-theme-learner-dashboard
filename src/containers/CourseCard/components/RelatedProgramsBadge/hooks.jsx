import React from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';

import { StrictDict } from 'utils';
import { hooks as appHooks } from 'data/redux';

import messages from './messages';
import * as module from './hooks';

export const state = StrictDict({
  isOpen: (val) => React.useState(val), // eslint-disable-line
});

export const useRelatedProgramsBadgeData = ({ cardId }) => {
  const [isOpen, setIsOpen] = module.state.isOpen(false);
  const { formatMessage } = useIntl();
  const numPrograms = appHooks.useCardRelatedProgramsData(cardId).length;
  let programsMessage = '';
  if (numPrograms) {
    programsMessage = formatMessage(
      numPrograms === 1 ? messages.badgeLabelSingular : messages.badgeLabelPlural,
      { numPrograms },
    );
  }

  return {
    numPrograms,
    programsMessage,
    isOpen,
    openModal: React.useCallback(() => setIsOpen(true), [setIsOpen]),
    closeModal: React.useCallback(() => setIsOpen(false), [setIsOpen]),
  };
};

export default useRelatedProgramsBadgeData;
