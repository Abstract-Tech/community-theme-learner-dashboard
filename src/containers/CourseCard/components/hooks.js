import { reduxHooks } from 'hooks';
import { EXECUTIVE_EDUCATION_COURSE_MODES } from '../../../data/constants/course';

export const useActionDisabledState = (cardId) => {
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const {
    canUpgrade, hasAccess, isAudit, isAuditAccessExpired, mode,
  } = reduxHooks.useCardEnrollmentData(cardId);
  const {
    isEntitlement, isFulfilled, canChange, hasSessions,
  } = reduxHooks.useCardEntitlementData(cardId);

  const { resumeUrl, homeUrl, upgradeUrl } = reduxHooks.useCardCourseRunData(cardId);

  const disableBeginCourse = !homeUrl || (isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired));
  const disableResumeCourse = !resumeUrl || (isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired));
  const disableViewCourse = !hasAccess || (isAudit && isAuditAccessExpired);
  const disableSelectSession = !isEntitlement || isMasquerading || !hasAccess || (!canChange || !hasSessions);
  const disableUpgradeCourse = !upgradeUrl || (isMasquerading && !canUpgrade);

  const disableCourseTitle = (isEntitlement && !isFulfilled) || disableViewCourse;

  const isExecutiveEd2uCourse = (EXECUTIVE_EDUCATION_COURSE_MODES.includes(mode));

  return {
    disableBeginCourse,
    disableResumeCourse,
    disableViewCourse,
    disableUpgradeCourse,
    disableSelectSession,
    disableCourseTitle,
    isExecutiveEd2uCourse,
  };
};

export default useActionDisabledState;
