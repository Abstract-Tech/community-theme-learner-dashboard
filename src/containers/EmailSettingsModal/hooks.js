import React from 'react';

import { StrictDict } from 'utils';
// import { thunkActions } from 'data/redux';
import { hooks as appHooks } from 'data/redux';

import * as module from './hooks';

export const state = StrictDict({
  toggle: (val) => React.useState(val), // eslint-disable-line
});

export const useEmailData = ({
  closeModal,
  courseNumber,
  // dispatch,
}) => {
  const { isEmailEnabled } = appHooks.useCardEnrollmentData(courseNumber);
  const [toggleValue, setToggleValue] = module.state.toggle(isEmailEnabled);
  const onToggle = React.useCallback(
    () => setToggleValue(!toggleValue),
    [setToggleValue, toggleValue],
  );
  const save = React.useCallback(
    () => {
      closeModal();
    },
    [closeModal],
  );

  return {
    onToggle,
    save,
    toggleValue,
  };
};

export default useEmailData;
