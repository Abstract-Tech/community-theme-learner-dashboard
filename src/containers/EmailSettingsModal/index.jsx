import React from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';

import {
  ActionRow,
  Button,
  Form,
  ModalDialog,
} from '@edx/paragon';

import { nullMethod } from 'hooks';
import shapes from 'data/services/lms/shapes';

import hooks from './hooks';
import messages from './messages';

export const EmailSettingsModal = ({
  closeModal,
  show,
  cardData,
}) => {
  const dispatch = useDispatch();
  const {
    toggleValue,
    onToggle,
    save,
  } = hooks({ dispatch, closeModal, cardData });
  const { formatMessage } = useIntl();

  return (
    <ModalDialog
      isOpen={show}
      onClose={nullMethod}
      hasCloseButton={false}
    >
      <div className="bg-white p-3 rounded shadow" style={{ textAlign: 'start' }}>
        <h4>{formatMessage(messages.header)}</h4>
        <Form.Switch checked={!toggleValue} onChange={onToggle}>
          {formatMessage(toggleValue ? messages.emailsOff : messages.emailsOn)}
        </Form.Switch>
        <p>{formatMessage(messages.description)}</p>
        <ActionRow>
          <Button variant="tertiary" onClick={closeModal}>
            {formatMessage(messages.nevermind)}
          </Button>
          <Button onClick={save}>{formatMessage(messages.save)}</Button>
        </ActionRow>
      </div>
    </ModalDialog>
  );
};
EmailSettingsModal.propTypes = {
  cardData: shapes.courseRunCardData.isRequired,
  closeModal: PropTypes.func.isRequired,
  show: PropTypes.bool.isRequired,
};

export default EmailSettingsModal;
