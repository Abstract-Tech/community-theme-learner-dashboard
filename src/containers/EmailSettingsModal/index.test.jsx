import React from 'react';
import { useDispatch } from 'react-redux';
import { shallow } from 'enzyme';

import hooks from './hooks';
import EmailSettingsModal from '.';

jest.mock('./hooks', () => ({
  __esModule: true,
  default: jest.fn(),
}));

const hookProps = {
  toggleValue: false,
  onToggle: jest.fn().mockName('hooks.onToggle'),
  save: jest.fn().mockName('hooks.save'),
};

const props = {
  closeModal: jest.fn().mockName('closeModal'),
  show: true,
  courseNumber: 'test-course-number',
};

const dispatch = useDispatch();

describe('EmailSettingsModal', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  describe('behavior', () => {
    beforeEach(() => {
      hooks.mockReturnValueOnce(hookProps);
      shallow(<EmailSettingsModal {...props} />);
    });
    it('calls hook w/ dispatch from redux hook, and closeModal, courseNumber from props', () => {
      expect(hooks).toHaveBeenCalledWith({
        closeModal: props.closeModal,
        dispatch,
        courseNumber: props.courseNumber,
      });
    });
  });
  describe('render', () => {
    test('snapshot: emails disabled, show: false', () => {
      hooks.mockReturnValueOnce(hookProps);
      expect(shallow(<EmailSettingsModal {...props} show={false} />)).toMatchSnapshot();
    });
    test('snapshot: emails disabled, show: true', () => {
      hooks.mockReturnValueOnce(hookProps);
      expect(shallow(<EmailSettingsModal {...props} />)).toMatchSnapshot();
    });
    test('snapshot: emails enabled, show: true', () => {
      hooks.mockReturnValueOnce({
        ...hookProps,
        toggleValue: true,
      });
      expect(shallow(<EmailSettingsModal {...props} />)).toMatchSnapshot();
    });
  });
});
