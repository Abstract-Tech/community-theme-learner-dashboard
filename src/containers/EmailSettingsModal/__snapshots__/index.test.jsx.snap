// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailSettingsModal render snapshot: emails disabled, show: false 1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={false}
  onClose={[MockFunction hooks.nullMethod]}
  title=""
>
  <div
    className="bg-white p-3 rounded shadow"
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    <h4>
      <formatMessage
        msg="Receive course emails?"
      />
    </h4>
    <Form.Switch
      checked={true}
      onChange={[MockFunction hooks.onToggle]}
    >
      <formatMessage
        msg="Course emails are on"
      />
    </Form.Switch>
    <p>
      <formatMessage
        msg="Course emailsi include important information about your course."
      />
    </p>
    <ActionRow>
      <Button
        onClick={[MockFunction closeModal]}
        variant="tertiary"
      >
        <formatMessage
          msg="Nevermind"
        />
      </Button>
      <Button
        onClick={[MockFunction hooks.save]}
      >
        <formatMessage
          msg="Save settings"
        />
      </Button>
    </ActionRow>
  </div>
</ModalDialog>
`;

exports[`EmailSettingsModal render snapshot: emails disabled, show: true 1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  title=""
>
  <div
    className="bg-white p-3 rounded shadow"
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    <h4>
      <formatMessage
        msg="Receive course emails?"
      />
    </h4>
    <Form.Switch
      checked={true}
      onChange={[MockFunction hooks.onToggle]}
    >
      <formatMessage
        msg="Course emails are on"
      />
    </Form.Switch>
    <p>
      <formatMessage
        msg="Course emailsi include important information about your course."
      />
    </p>
    <ActionRow>
      <Button
        onClick={[MockFunction closeModal]}
        variant="tertiary"
      >
        <formatMessage
          msg="Nevermind"
        />
      </Button>
      <Button
        onClick={[MockFunction hooks.save]}
      >
        <formatMessage
          msg="Save settings"
        />
      </Button>
    </ActionRow>
  </div>
</ModalDialog>
`;

exports[`EmailSettingsModal render snapshot: emails enabled, show: true 1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  title=""
>
  <div
    className="bg-white p-3 rounded shadow"
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    <h4>
      <formatMessage
        msg="Receive course emails?"
      />
    </h4>
    <Form.Switch
      checked={false}
      onChange={[MockFunction hooks.onToggle]}
    >
      <formatMessage
        msg="Course emails are off"
      />
    </Form.Switch>
    <p>
      <formatMessage
        msg="Course emailsi include important information about your course."
      />
    </p>
    <ActionRow>
      <Button
        onClick={[MockFunction closeModal]}
        variant="tertiary"
      >
        <formatMessage
          msg="Nevermind"
        />
      </Button>
      <Button
        onClick={[MockFunction hooks.save]}
      >
        <formatMessage
          msg="Save settings"
        />
      </Button>
    </ActionRow>
  </div>
</ModalDialog>
`;
